textbook/2_grouping.qmd

```diff
@@ -0,0 +1,328 @@
+
+The human brain can detect and process color differences quickly and subconsciously. Color is a *pre-attentive feature* which means that the brain distinguishes between colors even without conscious attention. This effect is even more pronounced when there is a high contrast between the colors, as is the case with complimentary colors like blue and orange.
+
+This is good news for grouping. By encoding our categorical variable as color, we can leverage the brain's ability to infer the groups in our data. For scatterplots, this works particularly well when you have a small number of clusters. Color can be easy added without altering the layout or structure of the data visualization. This makes it a powerful way to compare the exact values of data points between groups.
+
+::: {.panel-tabset}
+
+# Altair
+```{python}
+alt.Chart(simpsons_df).mark_circle().encode(
+  color='Group',
+  x=alt.X('X'),
+  y=alt.Y('Y'),
+)
+```
+
+# ggplot
+
+```{r}
+# Load necessary libraries
+library(ggplot2)
+simpsons_df <- read.csv("data/simpsons_paradox_data_l.csv")
+
+ggplot(simpsons_df, aes(x = X, y = Y, color = Group)) +
+  geom_point()
+```
+
+:::
+
+It is critical that each color be clearly distinct from the others, as too-similar colors can confuse the viewer and make differentiation difficult. Similarly, it is wise to limit the number of distinct categories represented by color: adding too many different colors overwhelms the visualization and makes them harder to distinguish.
+
+Accessibility is another key concern when using color. Colorblindness affects approximately 8% of men and 0.5% of women globally, with the most common variant being deuteranopia, which impairs the ability to distinguish between red and green hues. While Altair’s default color palette is designed to be color-blind friendly, ggplot’s is not.
+
+
+### Shape
+
+Shape can be used strategically in combination with color to provide an additional visual cue. This is particularly effective in contexts where colors may be difficult to distinguish. 
+
+A classic example of the power of this combination is the humble traffic light. At first glance, traffic lights seem poorly designed: they use the worst possible colors for people the most common form of colorblindess: red and green. However, in many cities, traffic lights are encoded with using both color and shape. The combination of these encoding channels, alongside the use of predictable y-axis positions for both 'Stop' and 'Go' signals creates a powerful effect that has undoubtedly prevented many accidents.
+
+![](/img/traffic.png)
+
+We can encode shapes aesthetics the same way as color, this can make it easier to distinguish the groups. 
+
+::: {.panel-tabset}
+
+# Altair
+```{python}
+alt.Chart(simpsons_df).mark_circle().encode(
+  shape='Group',
+  color='Group',
+  x=alt.X('X'),
+  y=alt.Y('Y'),
+)
+```
+
+# ggplot
+
+```{r}
+# Load necessary libraries
+library(ggplot2)
+simpsons_df <- read.csv("data/simpsons_paradox_data_l.csv")
```

Comment:
let's use `readr::read_csv` over `read.csv`, there are some cases where the latter causes issues that are hard to detect. In general, I wonder if we should read from the public URL here instead of the relative path, so that student can easily copy the code and run on their computers?

Reply:
have changed!